File: src/action_mux_cpp/src/generic_subscriber.cpp
// generic_subscriber.cpp
// This file implements a generic subscriber node using rclcpp's GenericSubscription API.
// It subscribes to a topic (specified via parameters) and prints out the size of each received serialized message.

#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
// Include the header for GenericSubscription
#include "rclcpp/generic_subscription.hpp"

class GenericSubscriber : public rclcpp::Node
{
public:
  GenericSubscriber() : Node("generic_subscriber")
  {
    // Declare parameters: topic name and message type.
    this->declare_parameter("topic_name", "input_topic");
    this->declare_parameter("msg_type", "std_msgs/msg/String");

    // Retrieve the parameters using the get<T>() method.
    std::string topic_name = this->get_parameter("topic_name").get_value<std::string>();
    std::string msg_type = this->get_parameter("msg_type").get_value<std::string>();

    // Create a generic subscription.
    // The callback accepts a shared pointer to a SerializedMessage.
    generic_subscription_ = this->create_generic_subscription(
      topic_name,
      msg_type,
      rclcpp::QoS(10),
      [this](std::shared_ptr<rclcpp::SerializedMessage> msg) {
        this->callback(msg);
      }
    );

    RCLCPP_INFO(this->get_logger(), "Generic subscription created on topic: %s with type: %s",
                topic_name.c_str(), msg_type.c_str());
  }

private:
  // The callback receives a serialized message.
  // We print out the size of the received message.
  void callback(std::shared_ptr<rclcpp::SerializedMessage> msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received serialized message of size: %zu bytes", msg->size());
  }

  rclcpp::GenericSubscription::SharedPtr generic_subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<GenericSubscriber>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/action_mux_py/action_mux_pkg/generic_subscriber.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import importlib


class GenericSubscriber(Node):
    def __init__(self):
        super().__init__('generic_subscriber')

        # Declare parameters to specify topic name and message type
        self.declare_parameter('topic_name', 'input_topic')
        self.declare_parameter('msg_type', 'std_msgs.msg.String')

        # Retrieve the parameters
        topic_name = self.get_parameter('topic_name').get_parameter_value().string_value
        msg_type_str = self.get_parameter('msg_type').get_parameter_value().string_value

        # Dynamically import the message type
        module_name, type_name = msg_type_str.rsplit('.', 1)
        module = importlib.import_module(module_name)
        msg_type = getattr(module, type_name)

        # Create the subscriber
        self.subscription = self.create_subscription(
            msg_type,
            topic_name,
            self.callback,
            10
        )

        self.get_logger().info(f'Subscribed to {topic_name} with type {msg_type_str}')

    def callback(self, msg):
        # Print the message and its type
        self.get_logger().info(f'Received message: {str(msg)}')
        self.get_logger().info(f'Message type: {str(type(msg))}')


def main(args=None):
    rclpy.init(args=args)
    node = GenericSubscriber()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
